### PR TITLE
Add age statistics

### LINKED ISSUE #22: Add a survival chart
Show the lifetime of systems in buckets like "1 day", "two days", "more than 7 days".

### PR DESCRIPTION
**Pull request by mdellweg**
fixes #22

### CODE REVIEW

File: pulpanalytics/management/commands/summarize.py [hunk 1]
```diff
@@ -23,14 +24,7 @@ def _find_very_first_date_to_summarize():
...
 
-        earliest_system_datetime = System.objects.order_by('created').first().created
-        first_summary_date = date(
-            year=earliest_system_datetime.year,
-            month=earliest_system_datetime.month,
-            day=earliest_system_datetime.day
-        )
-
-        return first_summary_date
+        return System.objects.order_by('created').first().created.date()
```
Comment (bmbouter) [inline pulpanalytics/management/commands/summarize.py:27]: :+1:

File: pulpanalytics/management/commands/summarize.py [hunk 2]
```diff
@@ -47,18 +41,24 @@ def _get_next_date_to_summarize(cls):
...
         online_workers_qs = OnlineWorkers.objects.filter(system__in=systems)
-        online_workers_stats = online_workers_qs.aggregate(Avg('processes'), Avg('hosts'))
+        try:
+            online_workers_stats = online_workers_qs.aggregate(Avg('processes'), Avg('hosts'))
 
-        summary.online_workers.processes__avg = online_workers_stats['processes__avg']
-        summary.online_workers.hosts__avg = online_workers_stats['hosts__avg']
+            summary.online_workers.processes__avg = online_workers_stats['processes__avg']
+            summary.online_workers.hosts__avg = online_workers_stats['hosts__avg']
+        except TypeError:
```
Comment (bmbouter) [inline pulpanalytics/management/commands/summarize.py:49]: What situation is this handling?

File: pulpanalytics/models.py [hunk 3]
```diff
@@ -6,7 +6,7 @@ class System(models.Model):
     created = models.DateTimeField(auto_now_add=True)
 
     def __str__(self):
-        return f"SystemID={self.id}, Created={self.created}"
+        return f"SystemID={self.system_id}, Created={self.created}"
```
Comment (bmbouter) [inline pulpanalytics/models.py:9]: :+1:

File: pulpanalytics/models.py [hunk 4]
```diff
@@ -27,7 +27,7 @@ class Meta:
         verbose_name_plural = "online content apps"
 
     def __str__(self):
-        return f"System={self.system.id}, Processes={self.processes}, Hosts={self.hosts}"
+        return f"System={self.system.system_id}, Processes={self.processes}, Hosts={self.hosts}"
```
Comment (bmbouter) [inline pulpanalytics/models.py:30]:
```suggestion
        return f"SystemID={self.system.system_id}, Processes={self.processes}, Hosts={self.hosts}"
```

I've made a real mess lol

File: pulpanalytics/models.py [hunk 5]
```diff
@@ -39,7 +39,7 @@ class Meta:
         verbose_name_plural = "online workers"
 
     def __str__(self):
-        return f"System={self.system.id}, Processes={self.processes}, Hosts={self.hosts}"
+        return f"System={self.system.system_id}, Processes={self.processes}, Hosts={self.hosts}"
```
Comment (bmbouter) [inline pulpanalytics/models.py:42]:
```suggestion
        return f"SystemID={self.system.system_id}, Processes={self.processes}, Hosts={self.hosts}"
```

**Comment by bmbouter** [comment]
For this strategy to work, we need to likely keep the original data around indefinitely. I think what will happen here is that when [raw data is deleted](https://github.com/pulp/analytics.pulpproject.org/blob/main/pulpanalytics/management/commands/summarize.py#L119-L127) the system age will max out at 2 weeks.

It's fully anonymous so I don't think we're doing anyone a big favor by keeping it around. You just can't run statistics like this otherwise and this is a very useful statistic.

File: pulpanalytics/management/commands/summarize.py (hunk 2 above)
Comment (mdellweg) [reply pulpanalytics/management/commands/summarize.py:49]: This is being graceful with me having bad database records. (Systems without any other information)

**Comment by mdellweg** [comment]
Actually i was happy to assume that each system at least 2 weeks old lives like for eternity and put them all in the same bucket.

On the other hand we could add a first_seen field which we needed to update on all incoming system records. That way we would get easier db queries and could keep the 2 weeks cleanup.

File: pulpanalytics/models.py (hunk 4 above)
Comment (mdellweg) [reply pulpanalytics/models.py:30]: This is for console access only to be fair.

**Comment by bmbouter** [comment]
I'm struggling with the visualization of this data. My test data shows the graph below which isn't very interesting because it only has 0-day systems (my test data doesn't reuse SystemIDs).

Screenshot from 2022-10-14 12-58-28

The question this helps us answer is: How long-lived are Pulp systems? To answer that I imaged we'd not have it be a timeseries and instead be, e.g. a bar chart with weeks that show the number of installations present N weeks ago, or a single line like a [survival analysis](https://duckduckgo.com/?t=ffab&q=survival+analysis+graph+example&iax=images&ia=images&iai=https%3A%2F%2Flifelines.readthedocs.io%2Fen%2Flatest%2F_images%2Fintro_survival_function.png) style chart.

This though has it also being a timeseries which suggests answering "How long-lived are Pulp systems a year ago verus today?". That doesn't seem like as interesting or important a question as the non-timeseries option. What do you think about this?

Also I'm a little unsure on what this data is really telling us as a project. Maybe it's to help identify user retention? That is one drawback of the charting we already have we could see a chart that goes up and to the right (project growth) but it's could be that we're actually just gaining more users than we're loosing... Any thoughts on what you really want to understand with this?

File: pulpanalytics/models.py (hunk 4 above)
Comment (bmbouter) [reply pulpanalytics/models.py]: It's for the django admin site.

File: pulpanalytics/management/commands/summarize.py (hunk 2 above)
Comment (bmbouter) [reply pulpanalytics/management/commands/summarize.py:49]: I'd prefer we not add code that silences a data quality problem. If for some reason the production site was in this situation I'd like us to find it via a log error. Can this be changed?

File: pulpanalytics/signals.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+from django.db.models.signals import pre_save
+from django.dispatch import receiver
+from django.utils import timezone
+
+from pulpanalytics.models import System
+
+
+@receiver(pre_save, sender=System)
+def _system_pre_save(sender, instance, raw, **kwargs):
+    if not raw:
```
Comment (bmbouter) [inline pulpanalytics/signals.py:10]: What does this if statement do, can a comment be added please?

Comment (mdellweg) [reply pulpanalytics/signals.py:10]: https://docs.djangoproject.com/en/4.1/ref/signals/#pre-save

**Comment by bmbouter** [comment]
Here's what I get when I use this code to summarize the last 14 days of production data:

image

I'm mentally struggling to understand what these are telling me.

**Comment by mdellweg** [comment]
image

Sure it's about how you present the data. But that second graph is telling me something: All the other graphs are strongly influenced by the ephemeral systems. I think, we want to filter the systems by `age > 2` in the daily summary to get more useful information.

**Comment by bmbouter** [comment]
These graphs look much better to me, thanks! I put this on the agenda for Thursday's working group. My hope is we can merge them there.

Since this creates the notion of `first_seen` as this operates over time and the DailySummary data has > 14 days in it what do you expect to happen to the upper and lower graphs?

Also can the upper graph have either an x-axis label added, e.g. "age of system" or make the title of the chart `Systems by Age` or `Age of Systems`? That one aspect I think is unclear to the untrained viewer (which in this case is me).

**Comment by mdellweg** [comment]
Added the label now. I expect the demography chart to grow with the oldest system (and at some point we may need to have a cutoff, but only time will tell). The lower graph however has the cutoff builtin (>=8 days is just the last bucket a system can be in.). It will grow with the number of collected `DailySummaries` as the plugin graphs do.

**Review by bmbouter — CHANGES_REQUESTED** [review]
How would you feel about moving the graphs to the bottom or to their own tab? If not would you be willing for me to reorganize them in a followup PR?

**Comment by mdellweg** [comment]
> How would you feel about moving the graphs to the bottom or to their own tab? If not would you be willing for me to reorganize them in a followup PR?

If it helps getting it merged...

**Review by bmbouter — APPROVED** [review]
Let's do this!

**Pull request merged**